### PR TITLE
[service-bus] Improving recovery and error handling

### PR DESCRIPTION
**Pull request by richardpark-msft**
Adding in a _lot_ more error handling code and retries in areas that were shown to be weak during Service Bus stress testing.

The biggest issues were mostly:
- Some errors weren't accounted for at all (for instance, credentials expiring seemed to fall through)
- Some errors are getting stringized (or aren't proper types) so there's some string matching code in the errors.go. Shouldn't be permanent, but we can recognize things fairly well with what's in there now.
- Some spots didn't have retries but really needed them. For safety, there are also some bugs we're working around where some frames appear to be arriving out of order in go-amqp. Things like that just result in a connection reset, which should recover it.

Also, mopped up some other problems:
- Fixed the naming of the `NewReceiver/NewProcessor` functions so they have ForQueue/ForSubscription in them
- Moved over a version of the stderrTracer from event-hubs. It's not _great_, but it's functional enough to work with the stress.go. It could use some work though, it's a bit hacked together.

### CODE REVIEW

File: sdk/messaging/azservicebus/client.go [hunk 1]
```diff
@@ -111,72 +116,123 @@ func newClientImpl(config clientConfig, options ...ClientOption) (*Client, error
...
-
 	if lastError != nil {
 		return fmt.Errorf("errors while closing links: %w", lastError)
 	}
 	return nil
 }
+
+func (client *Client) addCloseable(id uint64, closeable interface {
+	Close(ctx context.Context) error
+}) {
```
Comment (seankane-msft) [inline sdk/messaging/azservicebus/client.go:224]: nit: You use the `closeable` interface you have defined here in other spots. Why not define this interface as a `closeable` to reduce code?

File: sdk/messaging/azservicebus/client.go [hunk 2]
```diff
@@ -111,72 +116,123 @@ func newClientImpl(config clientConfig, options ...ClientOption) (*Client, error
 	return client, err
 }
 
-// NewProcessor creates a Processor, which allows you to receive messages from ServiceBus.
-func (client *Client) NewProcessor(options ...ProcessorOption) (*Processor, error) {
-	processor, err := newProcessor(client.namespace, options...)
+// NewProcessor creates a Processor for a queue.
+func (client *Client) NewProcessorForQueue(queue string, options ...ProcessorOption) (*Processor, error) {
+	options = append(options, ProcessorWithQueue(queue))
```
Comment (jhendrixMSFT) [inline sdk/messaging/azservicebus/client.go:121]: Do we still need this option even though queue is a param now?

File: sdk/messaging/azservicebus/client.go [hunk 3]
```diff
@@ -111,72 +116,123 @@ func newClientImpl(config clientConfig, options ...ClientOption) (*Client, error
...
 func (client *Client) Close(ctx context.Context) error {
 	var lastError error
 
+	var links []interface {
+		Close(ctx context.Context) error
+	}
+
 	client.linksMu.Lock()
 
 	for _, link := range client.links {
```
Comment (jhendrixMSFT) [inline sdk/messaging/azservicebus/client.go:203]: This can be simplified to `copy(links, client.links)`.  End result is the same.

File: sdk/messaging/azservicebus/client.go [hunk 4]
```diff
@@ -111,72 +116,123 @@ func newClientImpl(config clientConfig, options ...ClientOption) (*Client, error
...
+func (client *Client) addCloseable(id uint64, closeable interface {
+	Close(ctx context.Context) error
+}) {
+	client.linksMu.Lock()
+	client.links[id] = closeable
+	client.linksMu.Unlock()
+}
+
+func (client *Client) getCleanupForCloseable() (uint64, func()) {
+	id := atomic.AddUint64(&client.linkCounter, 1)
```
Comment (jhendrixMSFT) [inline sdk/messaging/azservicebus/client.go:231]: I don't see `linkCounter` being used elsewhere.

File: sdk/messaging/azservicebus/internal/amqpInterfaces.go [hunk 5]
```diff
@@ -5,7 +5,9 @@ package internal
 
 import (
 	"context"
+	"time"
 
+	"github.com/Azure/azure-amqp-common-go/v3/rpc"
```
Comment (jhendrixMSFT) [inline sdk/messaging/azservicebus/internal/amqpInterfaces.go:10]: This will go away eventually right?

File: sdk/messaging/azservicebus/stress/stress.go [hunk 6]
```diff
@@ -24,9 +26,15 @@ import (
 // | summarize Sum=sum(valueCount) by bin(timestamp, 30s), name
 // | render timechart
 
-var messagesSent int32
-var messagesReceived int32
-var exceptions int32
+type Stats struct {
```
Comment (jhendrixMSFT) [inline sdk/messaging/azservicebus/stress/stress.go:29]: This is all currently part of public surface area, is that intentional?

Comment (richardpark-msft) [reply sdk/messaging/azservicebus/stress/stress.go]: No, unexported

File: sdk/messaging/azservicebus/client.go (hunk 3 above)
Comment (richardpark-msft) [reply sdk/messaging/azservicebus/client.go:203]: In this case `client.links` is a map. Am I missing a way to make that work?

File: sdk/messaging/azservicebus/client.go (hunk 2 above)
Comment (richardpark-msft) [reply sdk/messaging/azservicebus/client.go]: I still have the option because, internally, it's still the easiest way for me to deal with it without having to expand to two functions everywhere. Might refactor later.

I did unexport it though, that was unintentional.

File: sdk/messaging/azservicebus/internal/amqpInterfaces.go (hunk 5 above)
Comment (richardpark-msft) [reply sdk/messaging/azservicebus/internal/amqpInterfaces.go:10]: Issue tracking this is here: https://github.com/Azure/azure-sdk-for-go/issues/15646

Is there urgency around this one though? I originally thought we could do this at any time.

File: sdk/messaging/azservicebus/client.go (hunk 4 above)
Comment (richardpark-msft) [reply sdk/messaging/azservicebus/client.go:231]: Yes, it's just used to provide a unique ID for inserting into our `client.links` map. I build a closure with the returned ID so we basically only deal with it through return values after this.

File: sdk/messaging/azservicebus/client.go (hunk 1 above)
Comment (richardpark-msft) [reply sdk/messaging/azservicebus/client.go]: Done.

File: sdk/messaging/azservicebus/client.go (hunk 3 above)
Comment (jhendrixMSFT) [reply sdk/messaging/azservicebus/client.go:203]: Ah I missed that.

File: sdk/messaging/azservicebus/internal/amqpInterfaces.go (hunk 5 above)
Comment (jhendrixMSFT) [reply sdk/messaging/azservicebus/internal/amqpInterfaces.go:10]: No just making sure it will happen.

File: sdk/messaging/azservicebus/messageSettler.go [hunk 7]
```diff
@@ -24,70 +39,79 @@ func (s *messageSettler) useManagementLink(m *ReceivedMessage, linkRevision uint
...
-	}
+	for retrier.Try(ctx) {
+		var receiver internal.AMQPReceiver
+		var mgmt internal.MgmtClient
+		var linkRevision uint64
+
+		_, receiver, mgmt, linkRevision, lastErr = s.links.Get(ctx)
+
+		if lastErr != nil {
+			_ = s.links.RecoverIfNeeded(ctx, lastErr)
```
Comment (jhendrixMSFT) [inline sdk/messaging/azservicebus/messageSettler.go:58]: If we always treat RecoverIfNeeded as no-fail perhaps it would be better to have it not return an error (I know that feels weird).

Comment (richardpark-msft) [reply sdk/messaging/azservicebus/messageSettler.go:58]: Yeah, I was thinking that too. I have post-beta-1 plans where I do want to make usage of it more, but not yet in this one.

File: sdk/messaging/azservicebus/messageBatch.go [hunk 8]
```diff
@@ -34,29 +34,18 @@ func newMessageBatch(maxBytes int) *MessageBatch {
...
-
-func (e errMessageTooLarge) NonRetriable()    {}
-func (e errMessageTooLarge) MessageTooLarge() {}
-
 // Add adds a message to the batch if the message will not exceed the max size of the batch
-// If the message is too large, an error of type 'ErrMessageTooLarge' will be returned.
-func (mb *MessageBatch) Add(m *Message) error {
+// This function will return:
+// (true, nil) if the message was added.
+// (false, nil) if the message was too large to fit into the batch.
```
Comment (jhendrixMSFT) [inline sdk/messaging/azservicebus/messageBatch.go:40]: If it's just that the batch is full I would reword this "if the batch is full". The current wording makes it read like if the message were smaller it would fit.

**Review by jhendrixMSFT — APPROVED** [review]
